### PR TITLE
Release 2.12.1

### PR DESCRIPTION
**Pull request by chengzhuzhang**
## Description

Prepare for releasing 2.12.1 before migrating to the new code base using xarray/xcdat

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by tomvothecoder — COMMENTED** [review]
My Initial code review

File: conda-env/dev-nompi.yml [hunk 1]
```diff
@@ -26,7 +26,9 @@ dependencies:
   - matplotlib-base
   - netcdf4
   - output_viewer >=1.3.0
-  - numpy >=1.23.0
+  - numpy >=2.0.0,<3.0.0
+  - shapely >=2.0.0,<3.0.0
+  - xarray >=2024.03.0
   - shapely >=2.0.0,<3.0.0
   - xarray >=2023.02.0
```
Comment (tomvothecoder) [inline conda-env/dev-nompi.yml:33]:
```suggestion
  - numpy >=2.0.0,<3.0.0
  - shapely >=2.0.0,<3.0.0
  - xarray >=2023.02.0
```

My fault, gave you incorrect suggestion. Merging these fixes.

File: conda-env/dev.yml [hunk 2]
```diff
@@ -21,7 +21,9 @@ dependencies:
   - matplotlib-base
   - netcdf4
   - output_viewer >=1.3.0
-  - numpy >=1.23.0
+  - numpy >=2.0.0,<3.0.0
+  - shapely >=2.0.0,<3.0.0
+  - xarray >=2024.03.0
   - shapely >=2.0.0,<3.0.0
   - xarray >=2023.02.0
```
Comment (tomvothecoder) [inline conda-env/dev.yml:28]:
```suggestion
  - numpy >=2.0.0,<3.0.0
  - shapely >=2.0.0,<3.0.0
  - xarray >=2023.02.0
```

File: conda-env/dev-nompi.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ dependencies:
   - matplotlib-base
   - netcdf4
   - output_viewer >=1.3.0
-  - numpy >=1.23.0
+  - numpy >=2.0.0,<3.0.0
   - shapely >=2.0.0,<3.0.0
   - xarray >=2023.02.0
```
Comment (tomvothecoder) [inline conda-env/dev-nompi.yml:31]:
```suggestion
  - xarray >=2024.03.0
```

File: conda-env/dev.yml [hunk 4]
```diff
@@ -21,7 +21,7 @@ dependencies:
   - matplotlib-base
   - netcdf4
   - output_viewer >=1.3.0
-  - numpy >=1.23.0
+  - numpy >=2.0.0,<3.0.0
   - shapely >=2.0.0,<3.0.0
   - xarray >=2023.02.0
```
Comment (tomvothecoder) [inline conda-env/dev.yml:26]:
```suggestion
  - xarray >=2024.03.0
```

**Comment by tomvothecoder** [comment]
Use this notebook for regression testing and modify as needed (e.g., update paths, remove function that removes `.nc` files for comparison, etc.) 

https://github.com/E3SM-Project/e3sm_diags/blob/cdat-migration-fy24/auxiliary_tools/cdat_regression_testing/843-migration-phase3/run-script-model-vs-obs/model_vs_obs_netcdf.ipynb

**Review by tomvothecoder — APPROVED** [review]
LGTM! Thanks Jill!

**Comment by chengzhuzhang** [comment]
@tomvothecoder Thank you for the review!
As a reference, the complete run is available here:
https://portal.nersc.gov/cfs/e3sm/chengzhu/complete_run/v2.12.1/e3sm_diags_extended/viewer/

The annual cycle zonal mean set is not included, only `ANN` is included in the run script.

**Comment by tomvothecoder** [comment]
Nice, thanks for the release @chengzhuzhang. Looks like we can merge `cdat-migration-fy24` to `main` soon 😀

**Comment by chengzhuzhang** [comment]
> Nice, thanks for the release @chengzhuzhang. Looks like we can merge `cdat-migration-fy24` to `main` soon 😀

YEAHHH!

**Pull request merged**